# U.S. Medical Insurance Costs
Here we will examine our Medical Insurance Data


1. First make sure you have pandas, tabulate, datetime, and plotly libraries installed.
2. We will first look at the data globally.
3. We will then look at the data by region.
4. After that we will break it down by generation, and number of children and finally by smoking status.

In [18]:
import pandas as pd
from tabulate import tabulate
import datetime
import plotly.express as px

df = pd.read_csv('insurance.csv')

# tabulet table format
tabul_format = 'pretty'  # https://pypi.org/project/tabulate/ -- go here for different table formats

print('The current date and time is: ', datetime.datetime.now())
print('===============================================================================================================')
print('Global Data')

# Getting general data from the dataset
# Create a global_list for the variables below to be appended to / created a header list for the tablulate table
global_list = []
header_list_global = ['Avg Age', 'Total Items', 'Total Smokers', 'Total Insurance Costs', 'Total BMI', 'Total Avg Costs', 'Total Avg BMI']

# Get my variables to be appended to the global_list
total_age_mean = round(df['age'].mean())
total_items = df['region'].count()
total_smokers = sum(df['smoker'] == 'yes')
total_insurance_costs_global = round(df['charges'].sum(),2)
total_bmi_global = round(df['bmi'].sum(), 2)
total_avg_cost = round((total_insurance_costs_global / total_items),2)
total_avg_bmi = round((total_bmi_global / total_items),2)

# Append my global variables to the global_list
global_list.append(total_age_mean)
global_list.append(total_items)
global_list.append(total_smokers)
global_list.append(total_insurance_costs_global)
global_list.append(total_bmi_global)
global_list.append(total_avg_cost)
global_list.append(total_avg_bmi)

# Had to make this so I could use tabulate for a single item list / had to create a list of lists
tabulate_end_list = ['x', 'x', 'x', 'x', 'x', 'x', 'x']
# Create a empty list to append the global list / tabulate_end_list to create the tabulate table
global_list_total = []
global_list_total.append(global_list)
global_list_total.append(tabulate_end_list)

# print out the tabulate table
print(tabulate(global_list_total, headers=header_list_global, tablefmt=tabul_format, floatfmt='.2f'))
print('===============================================================================================================')

The current date and time is:  2021-01-01 15:17:22.840701
Global Data
+---------+-------------+---------------+-----------------------+-----------+-----------------+---------------+
| Avg Age | Total Items | Total Smokers | Total Insurance Costs | Total BMI | Total Avg Costs | Total Avg BMI |
+---------+-------------+---------------+-----------------------+-----------+-----------------+---------------+
|   39    |    1338     |      274      |      17755824.99      | 41027.62  |    13270.42     |     30.66     |
|    x    |      x      |       x       |           x           |     x     |        x        |       x       |
+---------+-------------+---------------+-----------------------+-----------+-----------------+---------------+


From the data you can see the average age of the dataset it 39, average BMI is 30.66, and the average costs per individual is $13,270.  In the group there are 274 smokers, which represents a little over 20 percent of the dataset.

In [7]:
fig1 = px.bar(df, x='age', y='charges', color='smoker', barmode='group', title='Charges Per Age Group')
fig1.show()

From the chart above it appears that smoking significantly increases your insurance costs.  You can also see that the number of smokers decreases with age.  We will examine this further as we continue to dissect the data.

In [ ]:
results_dict = {}

# This function will allow us to get stats for any column value we want for example are they a smoker
def sort_by_col(column, value):
    data_col = column
    value_to_lookup = value
    # print('Global', value.capitalize(), 'Data')
    if data_col == 'age':
        print('Global', data_col.capitalize(), 'Data')
        dict_key = 'Global {} Data'.format(data_col.capitalize())
    elif data_col == 'sex' and value_to_lookup == 'female':
        print('Global Female Data')
        dict_key = 'Global Female Data'
    elif data_col == 'sex' and value_to_lookup == 'male':
        print('Global Male Data')
        dict_key = 'Global Male Data'
    elif data_col == 'children':
        print('Number of Children', value_to_lookup)
        dict_key = 'Number of Children {}'.format(value_to_lookup)
    elif data_col == 'smoker' and value_to_lookup == 'yes':
        print('Is a Smoker')
        dict_key = 'Is a Smoker'
    elif data_col == 'smoker' and value_to_lookup == 'no':
        print('Is not a Smoker')
        dict_key = 'Is not a Smoker'
    elif data_col == 'region':
        print('Global', value_to_lookup.capitalize(), 'Data')
        dict_key = 'Global {} Data'.format(value_to_lookup.capitalize())
    elif data_col == 'charges':
        print('Global Charges Data')
        dict_key = 'Global Charges Data'

    # Create a DataFrame for the item we want to inspect.
    new_df = df[(df[data_col] == value_to_lookup)]


# Create a global_list for the variables below to be appended to / created a header list for the tablulate table
    list_item = []
    header_list_global_smoker = ['Avg Age', 'Total Items', 'Total Smokers',
                             'Total Insurance Costs', 'Total BMI', 'Total Avg Costs', 'Total Avg BMI']

    # Get my variables to be appended to the global_list
    age_mean = round(new_df['age'].mean())
    items_total = new_df['region'].count()
    smoker_item_total = sum(new_df['smoker'] == 'yes')
    item_total_cost = round(new_df['charges'].sum(), 2)
    item_bmi_total = round(new_df['bmi'].sum(), 2)
    item_avg_cost = round((item_total_cost / items_total), 2)
    item_avg_bmi = round((item_bmi_total / items_total), 2)

    # Append my global variables to the global_list
    list_item.append(age_mean)
    list_item.append(items_total)
    list_item.append(smoker_item_total)
    list_item.append(item_total_cost)
    list_item.append(item_bmi_total)
    list_item.append(item_avg_cost)
    list_item.append(item_avg_bmi)

    # Had to make this so I could use tabulate for a single item list / had to create a list of lists
    tabulate_end_list_smoker = ['x', 'x', 'x', 'x', 'x', 'x', 'x']
    # Create a empty list to append the global list / tabulate_end_list to create the tabulate table
    tabulate_list = []
    tabulate_list.append(list_item)
    tabulate_list.append(tabulate_end_list_smoker)
    
    data_dict = {dict_key:[age_mean, items_total, smoker_item_total, item_total_cost, item_bmi_total, item_avg_cost, item_avg_bmi]}
    results_dict.update(data_dict)

    # print out the tabulate table
    print(tabulate(tabulate_list, headers=header_list_global_smoker, tablefmt=tabul_format, floatfmt='.2f'))
    print('===============================================================================================================')


In [11]:
sort_by_col('region', 'southwest')
sort_by_col('region', 'southeast')
sort_by_col('region', 'northwest')
sort_by_col('region', 'northeast')

Global Southwest Data
+---------+-------------+---------------+-----------------------+-----------+-----------------+---------------+
| Avg Age | Total Items | Total Smokers | Total Insurance Costs | Total BMI | Total Avg Costs | Total Avg BMI |
+---------+-------------+---------------+-----------------------+-----------+-----------------+---------------+
|   39    |     325     |      58       |      4012754.65       |  9943.9   |    12346.94     |     30.6      |
|    x    |      x      |       x       |           x           |     x     |        x        |       x       |
+---------+-------------+---------------+-----------------------+-----------+-----------------+---------------+
Global Southeast Data
+---------+-------------+---------------+-----------------------+-----------+-----------------+---------------+
| Avg Age | Total Items | Total Smokers | Total Insurance Costs | Total BMI | Total Avg Costs | Total Avg BMI |
+---------+-------------+---------------+-------------------

Here are some key observations when you break down the data by region:
1. The Southeast region has the highest average cost, total items, BMI, and number of smokers.
2. The Northeast region comes in second as it relates to total average costs and smokers.
3. As would be expected all four regions average age matches that of the global data.
4. It is obvious that if you are overweight and smoke your cost will be higher.

In [19]:
baby_boomers = [56,74]
gen_x = [40, 55]
millennials = [24,39]
gen_z = [8,23]

results_dict_age = {}

print('===============================================================================================================')

def age_function(generation, gen_name):
    print(gen_name + ' Data Age', generation[0], 'to', generation[1])
    age_range = df[df['age'].between(generation[0], generation[1], inclusive=True)]
    # Create a global_list for the variables below to be appended to / created a header list for the tablulate table
    list_item = []
    header_list_global_age = ['Avg Age', 'Total Items', 'Total Smokers',
                             'Total Insurance Costs', 'Total BMI', 'Total Avg Costs', 'Total Avg BMI']

    # Get my variables to be appended to the global_list
    age_mean_age = round(age_range['age'].mean())
    items_total_age = age_range['region'].count()
    age_item_total = sum(age_range['smoker'] == 'yes')
    item_total_cost_age = round(age_range['charges'].sum(), 2)
    item_bmi_total_age = round(age_range['bmi'].sum(), 2)
    item_avg_cost_age = round((item_total_cost_age / items_total_age), 2)
    item_avg_bmi_age = round((item_bmi_total_age / items_total_age), 2)

    # Append my global variables to the global_list
    list_item.append(age_mean_age)
    list_item.append(items_total_age)
    list_item.append(age_item_total)
    list_item.append(item_total_cost_age)
    list_item.append(item_bmi_total_age)
    list_item.append(item_avg_cost_age)
    list_item.append(item_avg_bmi_age)

    # Had to make this so I could use tabulate for a single item list / had to create a list of lists
    tabulate_end_list_age = ['x', 'x', 'x', 'x', 'x', 'x', 'x']
    # Create a empty list to append the global list / tabulate_end_list to create the tabulate table
    tabulate_list_age = []
    tabulate_list_age.append(list_item)
    tabulate_list_age.append(tabulate_end_list_age)
    
    data_dict_age = {gen_name:[age_mean_age, items_total_age, age_item_total, item_total_cost_age, item_bmi_total_age, item_avg_cost_age, item_avg_bmi_age]}
    results_dict_age.update(data_dict_age)

    # print out the tabulate table
    print(tabulate(tabulate_list_age, headers=header_list_global_age, tablefmt=tabul_format, floatfmt='.2f'))
    
    
    print('===============================================================================================================')

age_function(baby_boomers, 'Baby Boomer')    
age_function(gen_x, 'Generation X')
age_function(millennials, 'Millennial')
age_function(gen_z, 'Generation Z')

Baby Boomer Data Age 56 to 74
+---------+-------------+---------------+-----------------------+-----------+-----------------+---------------+
| Avg Age | Total Items | Total Smokers | Total Insurance Costs | Total BMI | Total Avg Costs | Total Avg BMI |
+---------+-------------+---------------+-----------------------+-----------+-----------------+---------------+
|   60    |     216     |      40       |      4059934.66       |  6855.7   |    18795.99     |     31.74     |
|    x    |      x      |       x       |           x           |     x     |        x        |       x       |
+---------+-------------+---------------+-----------------------+-----------+-----------------+---------------+
Generation X Data Age 40 to 55
+---------+-------------+---------------+-----------------------+-----------+-----------------+---------------+
| Avg Age | Total Items | Total Smokers | Total Insurance Costs | Total BMI | Total Avg Costs | Total Avg BMI |
+---------+-------------+---------------+--

When you break the data down into traditional generational brackets you can see that the Baby Boomer generation pays the most and has the lowest level of smokers.  This would show that age as related to costs is much more significant than other factors once you reach that level.  The data also shows that most of the dataset is in the Millennial and Generation X range and that after you turn 40 years old expect a significant increase in cost.  As we saw in the chart above insurance cost is reasonable for the Gen Z age group unless you smoke.  

In [16]:
sort_by_col('children', 0)
sort_by_col('children', 1)
sort_by_col('children', 2)
sort_by_col('children', 3)
sort_by_col('children', 4)
sort_by_col('children', 5)

Number of Children 0
+---------+-------------+---------------+-----------------------+-----------+-----------------+---------------+
| Avg Age | Total Items | Total Smokers | Total Insurance Costs | Total BMI | Total Avg Costs | Total Avg BMI |
+---------+-------------+---------------+-----------------------+-----------+-----------------+---------------+
|   38    |     574     |      115      |       7098070.0       | 17535.87  |    12365.98     |     30.55     |
|    x    |      x      |       x       |           x           |     x     |        x        |       x       |
+---------+-------------+---------------+-----------------------+-----------+-----------------+---------------+
Number of Children 1
+---------+-------------+---------------+-----------------------+-----------+-----------------+---------------+
| Avg Age | Total Items | Total Smokers | Total Insurance Costs | Total BMI | Total Avg Costs | Total Avg BMI |
+---------+-------------+---------------+---------------------

When you break the data down by number of children you will see that those with two and three children see the highest insurance costs.  This would be attributed to a higher average age and more smokers in the group proportionately.  As a side note I would think if you had 5 children it would drive you to smoking lol, but that doesn't appear to be the case.

In [17]:
sort_by_col('smoker', 'yes')
sort_by_col('smoker', 'no')


Is a Smoker
+---------+-------------+---------------+-----------------------+-----------+-----------------+---------------+
| Avg Age | Total Items | Total Smokers | Total Insurance Costs | Total BMI | Total Avg Costs | Total Avg BMI |
+---------+-------------+---------------+-----------------------+-----------+-----------------+---------------+
|   39    |     274     |      274      |      8781763.52       |  8414.12  |    32050.23     |     30.71     |
|    x    |      x      |       x       |           x           |     x     |        x        |       x       |
+---------+-------------+---------------+-----------------------+-----------+-----------------+---------------+
Is not a Smoker
+---------+-------------+---------------+-----------------------+-----------+-----------------+---------------+
| Avg Age | Total Items | Total Smokers | Total Insurance Costs | Total BMI | Total Avg Costs | Total Avg BMI |
+---------+-------------+---------------+-----------------------+-----------

This data is self-explanatory, and we have seen the trend in the dataset before. If you smoke your insurance costs are going to be significantly higher than those that don't smoke.

In conclusion one would say that the findings in this dataset make a lot of sense.  As you get older your insurance costs are going to increase.  This most significant jump is when you reach 40 years old.  The other two major factors are if you smoke and are you overweight.